### PR TITLE
ARTEMIS-4582 - view and update permissions, for jmx. and activemq.man…

### PR DESCRIPTION
**Pull request by gtully**
…agement. security-settings

### CODE REVIEW

**Comment by jbertram** [comment]
Couple of question...

1. One of the things that `management.xml` let you do was configure remote JMX connectivity. How do I do that with this?
2. One nice thing about `management.xml` is that you didn't have to muck around with system properties. Any chance this could be configured via `broker.xml` instead of `javax.management.builder.initial`? Perhaps a new `<management>` block would be useful here.
3. There is a fair bit of documentation about `management.xml` in `management.adoc`. Could you provide something analogous (or even more comprehensive) for this? A few simple use-cases with corresponding configuration would go a long way in helping users & developers understand how it works.
4. The existing default `management.xml` has recommended settings (e.g. read-only access to non-Artemis MBeans). I think we should have something similar for this as well. Is that possible?

**Comment by gtully** [comment]
> Couple of question...
> 
> 1. One of the things that `management.xml` let you do was configure remote JMX connectivity. How do I do that with this?
management.xml will remain, this will just replace the authorization section. 

> 2. One nice thing about `management.xml` is that you didn't have to muck around with system properties. Any chance this could be configured via `broker.xml` instead of `javax.management.builder.initial`? Perhaps a new `<management>` block would be useful here.

the system property is important because I want to guard the platform mbean server, and this is created on startup. Typically our logging causes it is be instantiated. When the platform mbean server is guarded and jolokia exposes it, the need for a jmx connector diminishes. With the management wrapper of the broker, we were late setting the guard in the past.  

> 3. There is a fair bit of documentation about `management.xml` in `management.adoc`. Could you provide something analogous (or even more comprehensive) for this? A few simple use-cases with corresponding configuration would go a long way in helping users & developers understand how it works.

sure. it is considerably simpler, with a mapping from objectName to address, and methods split into update or view permissions.

> 4. The existing default `management.xml` has recommended settings (e.g. read-only access to non-Artemis MBeans). I think we should have something similar for this as well. Is that possible?

that is a good idea, one thing that may get in the way is the default # security-settings-match. that will be applicable to jmx too but won't typically have the view role.

but a security-settings-match jmx.# view role would do it.

**Comment by gtully** [comment]
need to work on propagating the new permission into the security control apis, will revert draft when that is complete.

**Comment by gtully** [comment]
one challenge is keeping things simple, but also this needs to be independent. The VIEW and UPDATE permissions are applicable to both management messages on the activemq.management address and on MBean access from JMX and the console, where operations on particular MBeans should be locked down.

This means that I cannot  reuse security settings for addresses and queues and need to isolate the two entry points.

a postfix on the activemq.management address for management messages, such that a match of activemq.management.control.# will give full access.

and a prefix of jmx for mbean access, eg:
 jmx.broker.# // for all broker control operations
 jmx.addresses.activemq.management.pause // for the pause operation on a particular address

**Comment by gtully** [comment]
a question. the object name for an queue, Q1::Q1:

org.apache.activemq.artemis:broker="localhost",component=addresses,address="Q1",subcomponent=queues,routing-type="anycast",queue="Q1", 

maps nicely to:
 jmx.queues.Q1::Q1

But should it be `jmx.queue.` the management messages use the resource name  queue rather than queues. The MbeanObject name use attribute subcomponent=queues, queue=x. I guess it should be queue. No point polluting the settings with plurals when not necessary.

File: artemis-server/src/test/java/org/apache/activemq/artemis/core/security/RoleTest.java [hunk 1]
```diff
@@ -91,12 +91,12 @@ public void testManageRole() throws Exception {
...
-      Role roleWithDifferentWrite = new Role("testEquals", true, false, true, false, false, false, false, false, false, false);
-      Role roleWithDifferentCreate = new Role("testEquals", true, true, false, false, false, false, false, false, false, false);
+      Role role = new Role("testEquals", true, true, true, false, false, false, false, false, false, false, false, false);
+      Role sameRole = new Role("testEquals", true, true, true, false, false, false, false, false, false, false, false, false);
+      Role roleWithDifferentName = new Role("notEquals", true, true, true, false, false, false, false, false, false, false, false, false);
+      Role roleWithDifferentRead = new Role("testEquals", false, true, true, false, false, false, false, false, false, false, false, false);
+      Role roleWithDifferentWrite = new Role("testEquals", true, false, true, false, false, false, false, false, false, false, false, false);
+      Role roleWithDifferentCreate = new Role("testEquals", true, true, false, false, false, false, false, false, false, false, false, false);
```
Comment (gemmellr) [inline artemis-server/src/test/java/org/apache/activemq/artemis/core/security/RoleTest.java:99]: Seems like you should be adding some tests around here

**Review by gemmellr — COMMENTED** [review]
I added some comments on the earlier bits...on the later bits I've only skimmed a little of it, mostly I haven't yet looked at it as its just so vast and my head is spinning from hours of looking at it already...

**Review by gemmellr — COMMENTED** [review]
Gave the docs a read over, some more comments around clarity and gotchas etc.

File: artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/tools/xml/XmlDataImporter.java [hunk 2]
```diff
@@ -337,17 +337,27 @@ private void sendMessage(List<String> queues, Message message) throws Exception
...
                ClientMessage reply = requestor.request(managementMessage);
-               Number idObject = (Number) ManagementHelper.getResult(reply);
-               queueID = idObject.longValue();
+               if (ManagementHelper.hasOperationSucceeded(reply)) {
+                  Number idObject = (Number) ManagementHelper.getResult(reply);
+                  queueID = idObject.longValue();
+               } else {
+                  if (debugLog) {
+                     logger.debug("Failed to get ID for {} is: {}", queue, ManagementHelper.getResult(reply, String.class));
+                  }
```
[the remark this answers is not in the document]
Comment (gtully) [reply artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/tools/xml/XmlDataImporter.java:346]: good catch, thanks. will fix.

File: artemis-cli/src/test/java/org/apache/activemq/cli/test/ArtemisTest.java [hunk 3]
```diff
@@ -786,7 +786,7 @@ private void testProperReloadWhenAddingUserViaManagement(boolean basic) throws E
 
       try {
          activeMQServerControl.createAddress("myAddress", RoutingType.ANYCAST.toString());
-         activeMQServerControl.addSecuritySettings("myAddress", "myRole", "myRole", "myRole", "myRole", "myRole", "myRole", "myRole", "myRole", "myRole", "myRole");
+         activeMQServerControl.addSecuritySettings("myAddress", "myRole", "myRole", "myRole", "myRole", "myRole", "myRole", "myRole", "myRole", "myRole", "myRole", "", "");
```
[the remark this answers is not in the document]
Comment (gtully) [reply artemis-cli/src/test/java/org/apache/activemq/cli/test/ArtemisTest.java:789]: fair, will revert that change. thanks

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/management/ObjectNameBuilder.java [hunk 4]
```diff
@@ -161,7 +161,7 @@ private String getActiveMQServerName() {
       return String.format("%s:broker=%s", domain, (jmxUseBrokerName && brokerName != null) ? ObjectName.quote(brokerName) : "artemis");
    }
 
-   public ObjectName getManagementContextObjectName() throws Exception {
-      return ObjectName.getInstance(String.format("hawtio:type=security,area=jmx,name=ArtemisJMXSecurity"));
+   public ObjectName getSecurityObjectName() throws Exception {
```
[the remark this answers is not in the document]
Comment (gtully) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/management/ObjectNameBuilder.java:164]: probably, good catch, explains why it was not changed before now. I will add back the deprecated api

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/core/security/Role.java [hunk 5]
```diff
@@ -109,6 +115,22 @@ public Role(final String name,
...
+               final boolean createDurableQueue,
+               final boolean deleteDurableQueue,
+               final boolean createNonDurableQueue,
+               final boolean deleteNonDurableQueue,
+               final boolean manage,
+               final boolean browse,
+               final boolean createAddress,
+               final boolean deleteAddress,
+               final boolean view,
+               final boolean update) {
```
[the remark this answers is not in the document]
Comment (gtully) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/core/security/Role.java:133]: The view and update permissions are not checked by default, so the manage permission is sufficient as it is today. manage is very specific, it controls whether a message consumed from the management address is applied to the registered control objects or rejected. 
Only if the broker is configured to check for the presence of view or update for more fine grained rbac on the management address messages, then the user will need to configure those permissions.

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/core/security/Role.java [hunk 6]
```diff
@@ -247,6 +271,12 @@ public String toString() {
       if (browse) {
          stringReturn.append(" browse ");
       }
+      if (update) {
+         stringReturn.append(" update ");
+      }
+      if (view) {
+         stringReturn.append(" view ");
+      }
```
[the remark this answers is not in the document]
Comment (gtully) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/core/security/Role.java:279]: agree, will fix. thanks!

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/AddressControlImpl.java [hunk 7]
```diff
@@ -485,7 +473,6 @@ public String sendMessage(final Map<String, String> headers,
          try {
             return sendMessage(addressInfo.getName(), server, headers, type, body, durable, user, password, createMessageId);
          } catch (Exception e) {
-            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (gtully) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/AddressControlImpl.java:488]: i am assuming adding the e to the thrown exception was considered a leak, will use the logger. thanks

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/config/PersistedSecuritySetting.java [hunk 8]
```diff
@@ -77,7 +83,7 @@ public PersistedSecuritySetting(final String addressMatch,
                                    final String manageRoles,
                                    final String browseRoles,
                                    final String createAddressRoles,
-                                   final String deleteAddressRoles) {
+                                   final String deleteAddressRoles, String viewRoles, String updateRoles) {
```
[the remark this answers is not in the document]
Comment (gtully) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/config/PersistedSecuritySetting.java:86]: sure.

File: artemis-server/src/test/java/org/apache/activemq/artemis/core/security/RoleTest.java (hunk 1 above)
Comment (gtully) [reply artemis-server/src/test/java/org/apache/activemq/artemis/core/security/RoleTest.java:99]: good catch, thanks :-)

File: docs/user-manual/management.adoc [hunk 9]
```diff
@@ -250,8 +251,11 @@ It can be disabled by setting `jmx-management-enabled` to `false` in `broker.xml
 
 ==== Role Based Authorisation for JMX
 
-Although by default Artemis uses the Java Virtual Machine's `Platform MBeanServer` this is guarded using role based authorisation that leverages the broker's JAAS plugin support.
-This is configured via the `authorisation` element in the `management.xml` configuration file and can be used to restrict access to attributes and methods on MBeans.
+Artemis uses the Java Virtual Machine's `Platform MBeanServer` by default. This is guarded using role based authorisation that leverages the broker's JAAS plugin support.
+
+The RBAC used to restrict access to Mbeans and their operations can be configured in two ways, via security-settings in broker.xml, described in xref:management.adoc#jmx-authorisation-in-broker-xml[JMX authorisation in broker.xml], or via the `authorisation` element in the `management.xml` that is described below.
+
```
[the remark this answers is not in the document]
Comment (gtully) [reply docs/user-manual/management.adoc:257]: going with "one of two", thanks.

File: docs/user-manual/management.adoc [hunk 10]
```diff
@@ -362,6 +366,63 @@ The `*` access is the catch all for everything other method that isn't specifica
...
+An xref:management.adoc#artemis_rbac_mbean_server_guard[MBeanServer interceptor] that delegates to the broker security manager must be configured with a JVM system property that allows it to be added to all MBeanServers in the JVM.
+
+This is configured via a system property as follows:
+
+[,sh]
+----
+ java -Djavax.management.builder.initial=org.apache.activemq.artemis.core.server.management.ArtemisRbacMBeanServerBuilder
+----
+NOTE: When this property is provided, the authorization section of management.xml should be omitted as that depends on an alternative MBeanServer interceptor.
```
[the remark this answers is not in the document]
Comment (gtully) [reply docs/user-manual/management.adoc:384]: going with IMPORTANT, will see if there is some good place in the code to try and detect but I think in general, auto detection of invalid config is a rabbit hole. Maybe there is a use case for different MBean servers configured with different guards. That is possible but improbable.

File: docs/user-manual/security.adoc [hunk 11]
```diff
@@ -68,6 +68,14 @@ This permission allows the user to browse a queue bound to the matching address.
...
 This permission allows the user to invoke management operations by sending management messages to the management address.
 
+The following two permissions pertain to operations on the management apis of the broker. They split management operations into two sets, read only for `view`, and `update` for mutating operations. The split is controlled by a regular expression. Methods that match will require the `view` permission, all others require `update`. The regular expression can be modified through the configuration attribute `view-permission-method-match-pattern`. These permissions are applicable using a suffix on the management address, and with a `jmx.` prefix for MBean access.
+
+view::
+This permission allows access to a read-only subset of management operations.
+
+update::
+This permission allows access to the mutating management operations, any operation not in the `view` set.
+
```
[the remark this answers is not in the document]
Comment (gtully) [reply docs/user-manual/security.adoc:78]: thanks, linking in to other two sections.

File: docs/user-manual/security.adoc [hunk 12]
```diff
@@ -153,9 +161,22 @@ You can do this using the fully qualified queue name (i.e. FQQN) in the `match`
 </security-setting>
 ----
 
-NOTE: Wildcard matching doesn't work in conjuction with FQQN.
+NOTE: Wildcard matching doesn't work in conjunction with FQQN.
 The explicit goal of using FQQN here is to be _exact_.
 
+=== Applying view and update permissions
```
[the remark this answers is not in the document]
Comment (gtully) [reply docs/user-manual/security.adoc:167]: thanks!

File: docs/user-manual/management.adoc [hunk 13]
```diff
@@ -362,6 +366,63 @@ The `*` access is the catch all for everything other method that isn't specifica
...
+
+To ensure no user has permission to xref:management.adoc#force_failover[force a failover] using the broker (server control) MBean, use the following that defines the empty roles set for a particular mutating operation on the `broker` component:
+[,xml]
+----
+<security-setting match="jmx.broker.forceFailover">
+   <permission type="update" roles=""/>
+</security-setting>
+----
+
+NOTE: all attribute access `get|is[attribute]` maps to a method named getAttribute for RBAC purposes. All other operations require a literal match.
```
[the remark this answers is not in the document]
Comment (gtully) [reply docs/user-manual/management.adoc]: I have made it use a literal match for attributes and operations. so it is now consistent across both.

File: docs/user-manual/management.adoc [hunk 14]
```diff
@@ -585,7 +656,32 @@ This is also configured in broker.xml:
 </security-setting>
 ----
 
-=== Example
+==== Fine grained RBAC on management messages
+There is optional RBAC on the content of the management messages sent to the management address.
+
+RBAC is enabled by providing a value for the attribute xref:configuration-index.adoc#management_rbac_suffix[management-rbac-suffix].
+With a `management-rbac-suffix` of `.control.` more fine-grained permissions on the management address can be configured using that suffix appended to the management address.
```
[the remark this answers is not in the document]
Comment (gtully) [reply docs/user-manual/management.adoc]: cleared that up. using different prefixes (and prefixes for both) for match address for messages and for jmx helps clarify. they are independent. the jmx. matches handled by the mbean guard, the management messages handled by the session that receives them. the manage permission is still required and is unchanged.

**Comment by gtully** [comment]
new thought, the two prefixes for the security settings match, jmx and mgmt_msg will get in the way of doing filtering of aggregate operations like list* on a server control. Ideally your broker.listQueues query would only return what you can view. At the level of the control, we don't know if the call is from jmx or management messages. So a single prefix would be better.
Say management_ops, or `mops` as a default, and have a boolean to enable more fine grained access control for management messages, when true, it will look for the `mops` prefix.

**Comment by gtully** [comment]
And on update, there is a case for suggesting a publisher would need the update permission on an address to send a message, or a consumer could be updating  by acking. It is not clear that the update permission is about management.
Something like `edit` is less ambiguous, considering all the other perms are about messaging. 
I think i am replacing update with edit.

**Comment by gtully** [comment]
@jbertram I think this is in a happy place if you have some cycles to peek again. thanks

**Comment by cardamon** [comment]
FYI this introduces a breaking change that should probably be documented @ https://activemq.apache.org/components/artemis/documentation/latest/versions.html#upgrading-from-2-32-0

If you're embedding the broker, and using a custom `org.apache.activemq.artemis.spi.core.security.ActiveMQSecurityManager5` implementation, one must now either override `org.apache.activemq.artemis.spi.core.security.ActiveMQSecurityManager#getUserFromSubject` or ensure that `org.apache.activemq.artemis.spi.core.security.jaas.UserPrincipal` is used in the return value of the implementation of `org.apache.activemq.artemis.spi.core.security.ActiveMQSecurityManager5#authenticate`.

Not doing so will cause authentication failures (`AMQ222216`) that can be quite difficult to debug ...

**Pull request merged**